### DISCUSSION TITLE
when Dao.RunInTransaction() will fail?

### DISCUSSION THREAD

**Discussion by Oggy4th** [pocketbase/pocketbase#2564] [Q&A] [answered]
Can use `Dao.RunInTransaction()` inside for loop? When can't I use `Dao.RunInTransaction()`?

**Comment by ganigeorgiev** [comment]
> Can use Dao.RunInTransaction() inside for loop?

Yes.

> When can't I use Dao.RunInTransaction()?

I'm not sure that I understand the question. Could you provide a sample/pseudo code of what you are trying to do?

If you are asking when `RunInTransaction()` will return an error - when its callback returns an error.

**Reply by Oggy4th** [reply]
Below reasons are from Firebase

A transaction can fail for the following reasons:
- The transaction contains read operations after write operations. Read operations must always come before any write operations.
- The transaction read a document that was modified outside of the transaction. In this case, the transaction automatically runs again. The transaction is retried a finite number of times.
- The transaction exceeded the maximum request size of 10 MiB.


Is there any list like this one?

And also can I use for loop inside `Dao.RunInTransaction()`?

**Reply by ganigeorgiev** [reply, marked as answer]
There are no such rules in PocketBase. `RunInTransaction` executes a regular SQLite transaction. It is allowed to nest multiple `RunInTransaction` calls (_as long as you use each nested `txDao`_). They all will execute as one transaction.
The only limitation that eventually could occur is if you have mixed read and write operations inside the transaction as that may trigger `SQLITE_BUSY` errors on another concurrent db operation but we have auto fail/retry handling on app level in addition to the SQLite `busy_timeout` pragma so it should be very rare.

> And also can I use for loop inside Dao.RunInTransaction()?

Again, Yes, but without a code sample or at least some context I'm not really sure that I understand what you are trying to do.

Why do you need to execute `RunInTransaction` in a for loop in the first place?
Do you want only a single loop execution to be within a single transaction or the entire for loop to be wrapped in a transaction?

**Reply by Oggy4th** [reply]
I am using `RunInTransaction` a lot of different ways. Just make sure fail reasons.
